components/higher-order/with-state/index.js

```diff
@@ -0,0 +1,41 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, getWrapperDisplayName } from '@wordpress/element';
+
+/**
+ * A Higher Order Component used to provide and manage internal component state
+ * via props.
+ *
+ * @param  {?Object}   initialState Optional initial state of the component
+ * @return {Component}              Wrapped component
+ */
+function withState( initialState = {} ) {
+	return ( OriginalComponent ) => {
+		class WrappedComponent extends Component {
+			constructor() {
+				super( ...arguments );
+
+				this.setState = this.setState.bind( this );
+
+				this.state = initialState;
+			}
+
+			render() {
+				return (
+					<OriginalComponent
+						{ ...this.props }
+						{ ...this.state }
+						setState={ this.setState }
+					/>
+				);
+			}
+		}
+
+		WrappedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'withState' );
```

Comment:
In other places, I omitted `with` prefix for the display name. We should either update all existing `getWrapperDisplayName` calls or put `state` here instead. Both work for me.

Reply:
>We should either update all existing `getWrapperDisplayName` calls or put `state` here instead.

Updated in rebased 6fba36df84cb1bf0469b1e1effec8824752a0d21.